pace-util/pace/util/grid/helper.py

```diff
@@ -0,0 +1,513 @@
+    edge_s: FloatFieldI
+    edge_n: FloatFieldI
+
+
+@dataclasses.dataclass
+class VerticalGridData:
+    """
+    Terms defining the vertical grid.
+
+    Eulerian vertical grid is defined by p = ak + bk * p_ref
+    """
+
+    # TODO: make these non-optional, make FloatFieldK a true type and use it
+    ptop: float
+    ks: int
+    ak: Optional[Any] = None
+    bk: Optional[Any] = None
+    p_ref: Optional[Any] = None
+    """
+    reference pressure (Pa) used to define pressure at vertical interfaces,
+    where p = ak + bk * p_ref
+    ptop is the top of the atmosphere and ks is the lowest index (highest layer) for
+    which rayleigh friction
+
+    """
+
+
+@dataclasses.dataclass(frozen=True)
+class ContravariantGridData:
+    """
+    Grid variables used for converting vectors from covariant to
+    contravariant components.
+    """
+
+    cosa: FloatFieldIJ
+    cosa_u: FloatFieldIJ
+    cosa_v: FloatFieldIJ
+    cosa_s: FloatFieldIJ
+    sina_u: FloatFieldIJ
+    sina_v: FloatFieldIJ
+    rsina: FloatFieldIJ
+    rsin_u: FloatFieldIJ
+    rsin_v: FloatFieldIJ
+    rsin2: FloatFieldIJ
+
+
+@dataclasses.dataclass(frozen=True)
+class AngleGridData:
+    """
+    sin and cos of certain angles used in metric calculations.
+
+    Corresponds in the fortran code to sin_sg and cos_sg.
+    """
+
+    sin_sg1: FloatFieldIJ
+    sin_sg2: FloatFieldIJ
+    sin_sg3: FloatFieldIJ
+    sin_sg4: FloatFieldIJ
+    cos_sg1: FloatFieldIJ
+    cos_sg2: FloatFieldIJ
+    cos_sg3: FloatFieldIJ
+    cos_sg4: FloatFieldIJ
+
+
+class GridData:
+    # TODO: add docstrings to remaining properties
+
+    def __init__(
+        self,
+        horizontal_data: HorizontalGridData,
+        vertical_data: VerticalGridData,
+        contravariant_data: ContravariantGridData,
+        angle_data: AngleGridData,
+    ):
+        self._horizontal_data = horizontal_data
+        self._vertical_data = vertical_data
+        self._contravariant_data = contravariant_data
+        self._angle_data = angle_data
+
+    @classmethod
+    def new_from_metric_terms(cls, metric_terms: MetricTerms):
+        # TODO fix <Quantity>.storage mask for FieldI
+        shape = metric_terms.lon.data.shape
+        edge_n = utils.make_storage_data(
+            metric_terms.edge_n.data,
+            (shape[0],),
+            axis=0,
+            backend=metric_terms.edge_n.gt4py_backend,
+        )
+        edge_s = utils.make_storage_data(
+            metric_terms.edge_s.data,
+            (shape[0],),
+            axis=0,
+            backend=metric_terms.edge_s.gt4py_backend,
+        )
+        edge_e = utils.make_storage_data(
+            metric_terms.edge_e.data,
+            (1, shape[1]),
+            axis=1,
+            backend=metric_terms.edge_e.gt4py_backend,
+        )
+        edge_w = utils.make_storage_data(
+            metric_terms.edge_w.data,
+            (1, shape[1]),
+            axis=1,
+            backend=metric_terms.edge_w.gt4py_backend,
+        )
```

Comment:
Can you use `metric_terms.edge_n.storage` here? Why do we need `make_storage_data` from `dsl`? If it is not possible to convert the quantity `metric_terms.edge_w` into a storage using only that object, it feels like there's something wrong with `quantity.storage`.

Reply:
Good point. There was an issue with `edge_e` and `edge_w` having dimension [1, Y_INTERFACE_DIM] with (True, True, False) mask, which kept these from calling `.storage` directly. Added a `UNIT_X_DIM` dimension in Metric Terms to create these fields properly in the first place.